linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -1713,6 +1716,263 @@ private[optimizer] abstract class OptimizerCore(
     case _                 => false
   }
 
+  /** Tries to insert `valTree` in place of the (unique) occurrence of `valName` in `body`.
+   *
+   *  This function assumes that `valName` is used only once, and only inside
+   *  `body`. It does not assume that `valTree` or `body` are side-effect-free.
+   *
+   *  The replacement is done only if we can show that it will not affect
+   *  evaluation order. In practice, this means that we only replace if we find
+   *  the occurrence of `valName` in the first evaluation context of `body`.
+   *  In other words, we verify that all the expressions that will evaluate
+   *  before `valName` in `body` are pure.
+   *
+   *  We consider a `VarRef(y)` pure if `valTree` does not contain any
+   *  assignment to `y`.
+   *
+   *  For example, we can replace `x` in the following bodies:
+   *
+   *  {{{
+   *  x
+   *  x + e
+   *  x.foo(...)
+   *  x.f
+   *  e + x                 // if `e` is pure
+   *  e0.foo(...e1, x, ...) // if `e0` is pure and non-null, and the `...e1`s are pure
+   *  if (x) { ... } else { ... }
+   *  }}}
+   *
+   *  Why is this interesting? Mostly because of inlining.
+   *
+   *  Inlining tends to create many bindings for the receivers and arguments of
+   *  methods. We must do that to preserve evaluation order, and not to evaluate
+   *  them multiple times. However, very often, the receiver and arguments are
+   *  used exactly once in the inlined body, and in-order. Using this strategy,
+   *  we can take the right-hand-sides of the synthetic bindings and inline them
+   *  directly inside the body.
+   *
+   *  This in turn allows more trees to remain JS-level expressions, which means
+   *  that `FunctionEmitter` has to `unnest` less often, further reducing the
+   *  amount of temporary variables.
+   *
+   *  ---
+   *
+   *  We only call this function when the `minify` option is on. This is for two
+   *  reasons:
+   *
+   *  - it can be detrimental to debuggability, as even user-written `val`s can
+   *    disappear, and their right-hand-side be evaluated out-of-order compared
+   *    to the source code;
+   *  - it is non-linear, as we can perform several traversals of the same body,
+   *    if it follows a sequence of `VarDef`s that can each be successfully
+   *    inserted.
+   */
+  private def tryInsertAtFirstEvalContext(valName: LocalName, valTree: Tree, body: Tree): Option[Tree] = {
```

Comment:
Just to check my understanding: We have to do this in post transform for 2 reasons:
- We do not currently have general side-effect tracking in PreTransform.
- We need the usage counts, which we only get with the PreTransform pass (an alternative would be to do the inlining optimistically and cancel).

Reply:
Both of these points are true and contribute to this design choice, but it's more complicated than that. The biggest factor is that the decision of inlining an earlier val can depend on the ability to inline a later val.

Consider the following, common shape:
```scala
val x = f() // representing an expression with maybe side effects
val y = g() // same
h(x, y)
```
If we do this transformation in the forward direction, when we see `h(x, y)`---and assuming we know it is the only `x` already, somehow---, we would have to refuse it because it would cross the side effects of `g()`. If `g()` is known not to have side effects, we can inline it but then we get:
```scala
h(f(), y)
```
At that point, we cannot inline `y` because its evaluation would cross the side effects of `f()`.

With the implemented algorithm, which works backwards as a post transform, we can successfully inline both `x` and `y`, even if both `f()` and `g()` have side effects. First we inline `y`, giving
```scala
val x = f()
h(x, g())
```
and then we can still inline `x`, giving
```scala
h(f(), g())
```
which is optimal.

No other strategy that I could think of would be able to deal with these situations. It must be a backwards decision AFAICT.

Reply:
Thanks for the explanation. It makes a lot of sense.

I cannot shake the feeling that we can get rid of the quadratic behavior with something-something dynamic programming, but from the example above, it is also clear that this is not trivial. So I think we should leave it for later.